Add MenuButton

alxHenry opened the pull request:
- Add the MenuButton component
- Add the MenuButtonIcon component that we pass to fabric as a custom renderer for menu items
- For now the only type of target will be an icon. In the future there will be a followup to add a button as a target for the menu.
- Fix a small existing issue that styleguidist was throwing console errors for
- Expose some additional existing text colors

## Pull request checklist

* [x] Component `README.md` file is up-to-date.
* [x] Component is unit tested.

alxHenry commented:
@swese44 @unindented we can start taking a look at the code. Once the fabric release is cut (should be tomorrow) I can do the visual diff stuff.

alxHenry commented:
@mrfunkycold

alxHenry commented:
@unindented @swese44 I bumped Fabric to the version with my changes and everything still looks good here and in Dropdown.

Added the initial visual diffs as well. Unfortunately fabric doesn't have a 'render with menu open' or anything. @unindented what is the repo for the screenshotter you made? Might take a look at the robo click of elements.

unindented commented:
@alxHenry https://github.com/unindented/react-styleguidist-visual

alxHenry commented:
@swese44 updated the tests, completed coverage, added the coloring of the icons

swese44 commented on src/components/MenuButton/MenuButton.css:
```diff
@@ -0,0 +1,58 @@
+}
+
+.y-menu-button--item {
+  padding: .4rem 1.2rem .8rem
```
semicolon at end of line

swese44 commented on src/components/MenuButton/MenuButton.css:
```diff
@@ -0,0 +1,58 @@
+  padding: .4rem 1.2rem .8rem
+}
+.y-menu-button--item-header {
+  padding: .6rem 1.2rem 1rem
```
semicolon at end of line

swese44 commented on src/components/MenuButton/MenuButton.css:
```diff
@@ -0,0 +1,58 @@
+}
+
+
+.y-menu-button--callout a:hover {
```
we should replace these `a` tag selectors with the classname of the `a` and `button` tags. also want to make sure the bgcolors are correctly the same on both tags

alxHenry replied:
These a tags have the same class as all the items, .ms-ContextualMenu-link. I replaced the a selector with that class and I think it should be okay.

alxHenry replied:
bgcolors look the same for link and non link

swese44 commented on src/components/MenuButton/MenuButton.md:
```diff
@@ -0,0 +1,76 @@
+      type: 2,
+    }, {
+      key: 'divider',
+      type: 1,
```
are we unable to use the enum here in the .md file instead of the compiled int value?

swese44 commented on src/components/MenuButton/MenuButton.md:
```diff
@@ -0,0 +1,76 @@
+    }, {
+      key: 'noIcon',
+      text: 'No Icon',
+      onClick: () => console.log('no icon'),
```
we should replace these `console.log()` statements with `action()`, other component example files should have this

swese44 commented on src/components/MenuButton/MenuButton.md:
````diff
@@ -0,0 +1,76 @@
+
+MenuButton with specified Icon and basic children
+
+```js { "props": { "data-description": "specified icon and icon size" } }
````
let's fix the description, maybe just "custom icon"?

swese44 replied:
you'll probably have to delete these screenshots from `config/styleguide-visual/reference` fyi

swese44 commented on src/components/MenuButton/MenuButton.md:
````diff
@@ -0,0 +1,76 @@
+
+MenuButton with default Icon, and all applicable children properties
+
+```js { "props": { "data-description": "with deafult icon and children properties" } }
````
"default" is misspelled.

how about "with default icon and all menu item types"

swese44 commented on src/components/MenuButton/MenuButton.tsx:
```diff
@@ -0,0 +1,89 @@
+export interface MenuButtonProps extends BaseComponentProps {
+  ariaLabel: string;
+  menuItems: MenuButtonItem[];
+  className?: string;
```
we can delete className, it gets inherited from BaseComponentProps

swese44 commented on src/components/MenuButton/MenuButton.tsx:
```diff
@@ -0,0 +1,89 @@
+const renderEmptyIcon = () => null;
+
+export interface MenuButtonItem {
+  key: string;
```
We should add comments before each item in the `MenuButtonItem` and `MenuButtonProps` interfaces, this will show up in the editor as TypeScript intellisense descriptions

adrianmg commented:
I've made some small adjustments here: https://github.com/Microsoft/YamUI/pull/254/commits/b569740e14b94832b847eacad85d64bee8226318

Sharing some questions:
- Could we generate more examples with different combinations with the titles and no titles? So I can check all the different options and margins.
- Is someone taking care of the document list case?

Pending things to be fixed after my commit:
- Row with icon: Move icon 1px down VS label
- Border radius is not applied to the general container (or not visible)
- The disabled label looks bolded instead of regular (this thicker even if the opacity is lower)
- Header shouldn't have 2 line dividers (check design spec)
- Why are we hiding the component whenever I press Alt/Cmd? I'm wondering if that's an accessibility issue since I may want to Cmd+Click to open a section in a new tab and that's preventing me from doing that. @BrendanMcK 
- The default is not the one we defined, I'm actually thinking on following this (re-open and check Zeplin small update on this):
- 1. Default state: defined by CSS
- 2. Hover: defined by CSS
- 3. Active: keep the hover state and force 100% opacity (unless something else custom CSS is defined ad-hoc)

The rest looks OK 👍

adrianmg commented:
@alxHenry clarifying some of those questions!

- Generating more cases (image attached): Having the header in different places so make sure the margins are always kept and flowing right with the vertical rhythm could be helpful.
- +1 creating an issue ticket for us/Fabric to fix the concerns regarding accessibility.
- I'm happy with defining '•••' as the default one.
- Regarding states: currently when the icon is active it loses opacity, my idea is to keep the same styling as the hover state and make sure opacity is 100% as the default behavior. All the three states (default, hover and active) should be easy to override if needed at some point in CSS.

![image](https://user-images.githubusercontent.com/589285/36813827-9fad541e-1c8a-11e8-8925-a976fc599416.png)

alxHenry commented:
@adrianmg are you referring to the icon being blue when the element is disabled?

I'm not seeing the icon ever have less than 100% opacity with the active state. Am I reading that right?

adrianmg commented:
@alxHenry sorry! Yes, I thought we were reducing the opacity because it seemed a lighter blue compared to the hover state, sorry!

alxHenry commented:
No worries, will fix that color

alxHenry commented:
@adrianmg I changed the examples a bit to have headers and dividers and both to test out cases.  Also updated the item icon to be gray when disabled.

I think that's everything, let me know if it isn't.

swese44 commented on src/components/MenuButton/MenuButtonItem.tsx:
```diff
@@ -0,0 +1,55 @@
+  const { data, name, itemType, disabled } = props.item;
+
+  const Icon = data && data.yamUIIcon;
+  const iconColumn = Icon ? (
```
it might be a little cleaner to move this to a private `getIconColumn()` method

unindented commented on src/components/MenuButton/MenuButton.css:
```diff
@@ -0,0 +1,80 @@
+  opacity: 1;
+}
+
+.y-menu-button--callout {
```
I think this is the opposite of BEM. If we were following BEM, this would be `y-menu-button__callout`.

unindented replied:
http://getbem.com/naming/

adrianmg commented:
@alxHenry disabled rows shouldn't be interactive (it currently shows an hover state).
